File: TypeScript/app/gilded-rose.ts
export class Item {
    name: string;
    sellIn: number;
    quality: number;

    constructor(name, sellIn, quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }
}

export class GildedRose {
    items: Array<Item>;

    constructor(items = [] as Array<Item>) {
        this.items = items;
    }

    updateQuality() {
        for (let item of this.items) {
            if (item.name.includes('Sulfuras')){
                item.quality = 80
            } else if (item.name.includes('Aged Brie') || item.name.includes('Backstage passes ')) {
                item.quality = item.quality + 1
                if (item.name.includes('Backstage passes')) {
                    if ( item.sellIn < 11 && item.sellIn > 6) {
                        item.quality = item.quality + 1
                    } else if (item.sellIn < 6 && item.sellIn >= 0) {
                        item.quality = item.quality + 2
                    } else if (item.sellIn < 0) {
                        item.quality = 0
                    }
                }
                if (item.quality > 50) {
                    item.quality = 50
                }
                item.sellIn = item.sellIn - 1;
            } else {
                if (item.quality > 0 && item.quality <= 50) {
                    if (item.sellIn < 0 || item.name.includes('Conjured')) {
                        item.quality = item.quality - item.quality
                    } else {
                        item.quality = item.quality - 1
                    }
                } else if (item.quality < 0) {
                    item.quality = 0
                } else if (item.quality > 50){
                    item.quality = 50
                }
                item.sellIn -= 1
            }
        }

        return this.items;
    }
}

File: python/gilded_rose.py
# -*- coding: utf-8 -*-


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)


class GildedRose(object):
    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            if "Sulfuras" in item.name:
                item.quality = 80
            elif "Brie" in item.name or "Backstage passes" in item.name:
                item.quality += 1
                if "Backstage passes" in item.name:
                    if item.sell_in < 11 and item.sell_in > 6:
                        item.quality += 1
                    elif item.sell_in < 6 and item.sell_in >= 0:
                        item.quality += 2
                    elif item.sell_in < 0:
                        item.quality = 0

                if item.quality > 50:
                    item.quality = 50

                item.sell_in -= 1
            else:
                if item.quality > 0 and item.quality <= 50:
                    if item.sell_in < 0 or "Conjured" in item.name:
                        item.quality = item.quality - item.quality
                    else:
                        item.quality -= 1
                elif item.quality < 0:
                    item.quality = 0
                elif item.quality > 50:
                    item.quality = 50

                item.sell_in -= 1
